File: backend/src/service/treehole.service.ts
import Treehole from '@/models/Treehole'
import { prisma } from '@/config/db'

export async function getMessage(limit: number) {
    const filter = { deleted: { $ne: true }, reviewStatus: 'APPROVED' }

    const danmakus = await Treehole.find(filter)
        .sort({ createTime: -1 })  // 按时间倒序，最新的在前
        .select('-reviewStatus -createTime -deleted')
        .limit(limit)              // 限制数量
        .lean()

    const userIds = [...new Set(danmakus.map(a => a.userId).filter(Boolean))];

    const users = await prisma.user.findMany({
        where: {
            user_id: {
                in: userIds
            },
            deleted: 0,
            account_status: 'ACTIVE'
        },
        select: {
            user_id: true,
            username: true,  // 只查询需要的字段
            cover: true
        }
    })

    const usersMap = new Map(users.map(c => [c.user_id, c.cover]));

    const formattedDanmakus = danmakus
        .map(danmaku => ({
            ...danmaku,
            avatar: usersMap.get(danmaku.userId) || '',
        }));

    return formattedDanmakus
}

// 检查用户是否发送太频繁
export async function checkRecentMessage(userId: number): Promise<boolean> {
    const query = { deleted: { $ne: true }, reviewStatus: 'APPROVED' };

    const recentMessage = await Treehole.findOne({
        userId,
        createTime: { $gt: new Date(Date.now() - 10 * 1000) },
        ...query
    })
    return !!recentMessage
}

// 创建新的树洞消息
export async function createMessage(content: string, userId: number) {
    await Treehole.create({
        content: content.trim(),
        userId,
        createTime: new Date()
    })
}

File: backend/java-backend/src/main/java/com/example/demo/service/TreeholeService.java
package com.example.demo.service;

import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.example.demo.common.BusinessException;
import com.example.demo.mapper.UserMapper;
import com.example.demo.model.mongo.Treehole;
import com.example.demo.model.mysql.User;
import com.example.demo.repository.mongo.TreeholeRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class TreeholeService {

    @Autowired
    private TreeholeRepository treeholeRepository;

    @Autowired
    private UserMapper userMapper;

    /** 获取最新的树洞消息列表，附带用户头像 */
    public List<Map<String, Object>> getMessage(int limit) {
        PageRequest pageRequest = PageRequest.of(0, limit, Sort.by(Sort.Direction.DESC, "createTime"));
        List<Treehole> messages = treeholeRepository.findByDeletedNotAndReviewStatusOrderByCreateTimeDesc(true, "APPROVED", pageRequest);

        // 批量获取用户头像
        Set<Integer> userIds = messages.stream()
                .map(Treehole::getUserId)
                .filter(Objects::nonNull)
                .collect(Collectors.toSet());
        Map<Integer, User> userMap = resolveUsers(userIds);

        List<Map<String, Object>> result = new ArrayList<>();
        for (Treehole msg : messages) {
            Map<String, Object> item = new LinkedHashMap<>();
            item.put("id", msg.getId());
            item.put("content", msg.getContent());
            item.put("createTime", msg.getCreateTime());
            item.put("userId", msg.getUserId());
            User user = userMap.get(msg.getUserId());
            if (user != null) {
                item.put("cover", user.getCover());
            }
            result.add(item);
        }

        return result;
    }

    /** 校验用户10秒内是否发过消息，防止刷屏 */
    public void checkRecentMessage(Integer userId) {
        LocalDateTime tenSecondsAgo = LocalDateTime.now().minusSeconds(10);
        boolean recent = treeholeRepository.existsByUserIdAndCreateTimeAfterAndDeletedNotAndReviewStatus(
                userId, tenSecondsAgo, true, "APPROVED");
        if (recent) {
            throw new BusinessException(429, "发送过于频繁，请稍后再试");
        }
    }

    /** 创建树洞消息，校验频率后保存为待审核状态 */
    public void createMessage(String content, Integer userId) {
        checkRecentMessage(userId);

        Treehole message = new Treehole();
        message.setContent(content);
        message.setUserId(userId);
        message.setCreateTime(LocalDateTime.now());
        message.setReviewStatus("PENDING");
        message.setDeleted(false);
        treeholeRepository.save(message);
    }

    /** 批量根据用户ID查询用户信息映射 */
    private Map<Integer, User> resolveUsers(Set<Integer> userIds) {
        if (userIds == null || userIds.isEmpty()) {
            return Collections.emptyMap();
        }
        List<User> users = userMapper.selectList(
                new LambdaQueryWrapper<User>().in(User::getUserId, userIds)
        );
        return users.stream().collect(Collectors.toMap(User::getUserId, u -> u, (a, b) -> a));
    }
}